File: spacetimedb/src/world/spawn.ts
import { Identity } from 'spacetimedb';
import {
  WORLD_SIZE,
  START_PEASANTS,
  START_WOOD,
  START_STONE,
  START_FOOD,
  START_GOLD,
  SPAWN_CLUSTER,
  MAX_PLAYERS,
} from '../../../shared/constants.ts';
import {
  UNIT_DEFS,
  BUILDING_DEFS,
  PLAYER_COLORS,
  AI_PROFILES,
  NODE_KINDS,
  type NodeKindDef,
  aiName,
  spawnCorner,
  allocSlot,
} from '../../../shared/defs.ts';
import { sampleTerrain, isLand, isCoastal } from '../../../shared/terrain.ts';
import { findBuildableNear } from '../../../shared/buildings.ts';
import {
  UnitKind,
  BuildingKind,
  ResourceType,
  GatherState,
  Faction,
  Stance,
  type UnitKind as UnitKindT,
  type BuildingKind as BuildingKindT,
} from '../../../shared/enums.ts';
import { clampWorld, getSeed, buildNodes, assignGather } from './util.ts';

// ctx is typed inside reducers; helpers take `any` to avoid threading the schema
// generic everywhere. They only touch typed table rows.
export function spawnUnitEntity(
  ctx: any,
  owner: any,
  kind: number,
  x: number,
  y: number
): bigint {
  const def = UNIT_DEFS[kind as UnitKindT] ?? UNIT_DEFS[UnitKind.Peasant];
  const e = ctx.db.entity.insert({ entityId: 0n, x, y, facing: 0 });
  ctx.db.unit.insert({
    entityId: e.entityId,
    owner,
    kind,
    targetX: x,
    targetY: y,
    hasTarget: false,
    speed: def.speed,
    gatherState: GatherState.Idle,
    targetNode: 0n,
    carrying: 0,
    carryType: ResourceType.Wood,
    harvestTimer: 0,
    hp: def.maxHp,
    attackTarget: 0n,
    attackCooldown: 0,
    stance: Stance.Aggressive,
    homeX: x,
    homeY: y,
    path: [],
    pathIdx: 0,
  });
  return e.entityId;
}

export function spawnBuilding(
  ctx: any,
  owner: any,
  kind: number,
  x: number,
  y: number
): bigint {
  const def = BUILDING_DEFS[kind as BuildingKindT] ?? BUILDING_DEFS[BuildingKind.Keep];
  const e = ctx.db.entity.insert({ entityId: 0n, x, y, facing: 0 });
  ctx.db.building.insert({
    entityId: e.entityId,
    owner,
    kind,
    hp: def.maxHp,
    cooldown: 0,
    rallyX: x,
    rallyY: y,
  });
  return e.entityId;
}

function spawnNode(
  ctx: any,
  x: number,
  y: number,
  resType: number,
  remaining: number
): void {
  const e = ctx.db.entity.insert({ entityId: 0n, x, y, facing: 0 });
  ctx.db.resourceNode.insert({ entityId: e.entityId, resType, remaining });
}

// Rejection-sample one node kind across the map by its per-biome density. Nodes
// only land on reachable ground: coastal-only kinds (fish) hug the shore, every
// other kind sits on passable land so a harvester can always reach it.
function scatterKind(ctx: any, seed: number, def: NodeKindDef): void {
  let placed = 0;
  let attempts = 0;
  const budget = Math.max(60, def.count) * 80;
  while (placed < def.count && attempts < budget) {
    attempts++;
    const x = 3 + ctx.random() * (WORLD_SIZE - 6);
    const y = 3 + ctx.random() * (WORLD_SIZE - 6);
    const reachable = def.coastalOnly
      ? isCoastal(seed, x, y)
      : isLand(seed, x, y);
    if (!reachable) continue;
    if (ctx.random() < def.density(sampleTerrain(seed, x, y).biome)) {
      spawnNode(ctx, x, y, def.resType, def.yield);
      placed++;
    }
  }
}

// Scatter every resource node kind across the map. Shared by init and match
// reset. Data-driven from NODE_KINDS — new node kinds need no code here.
export function scatterNodes(ctx: any, seed: number): void {
  for (const def of NODE_KINDS) scatterKind(ctx, seed, def);
}

// Found a new base for `owner`: keep at the next free corner, starting peasants
// already gathering. Shared by enterGame (human) and addAi (skirmish bot) so both
// sides start identically. Returns the keep entity id.
export function foundPlayer(
  ctx: any,
  owner: any,
  name: string,
  faction: number
): bigint {
  // Stable slot from the set of slots in use — survives leavers, so two players
  // can never share a corner (overlapping keeps). Caller guards MAX_PLAYERS.
  const used = [...ctx.db.player.iter()].map((p: any) => p.slot);
  const slot = Math.max(0, allocSlot(used, MAX_PLAYERS));
  const seed = getSeed(ctx);
  const corner = spawnCorner(slot);
  const base = findBuildableNear(
    seed,
    corner.x,
    corner.y,
    BUILDING_DEFS[BuildingKind.Keep].footprint
  );
  const keepId = spawnBuilding(ctx, owner, BuildingKind.Keep, base.x, base.y);

  ctx.db.player.insert({
    identity: owner,
    playerId: 0,
    name,
    faction,
    wood: START_WOOD,
    stone: START_STONE,
    food: START_FOOD,
    gold: START_GOLD,
    color: slot % PLAYER_COLORS.length,
    online: true,
    keepEntity: keepId,
    defeated: false,
    slot,
  });

  const nodes = buildNodes(ctx);
  for (let i = 0; i < START_PEASANTS; i++) {
    const a = (i / START_PEASANTS) * Math.PI * 2;
    const px = clampWorld(base.x + Math.cos(a) * SPAWN_CLUSTER);
    const py = clampWorld(base.y + Math.sin(a) * SPAWN_CLUSTER);
    const id = spawnUnitEntity(ctx, owner, UnitKind.Peasant, px, py);
    assignGather(ctx, id, px, py, nodes);
  }
  return keepId;
}

// Spawn one bot for `host`'s match. Identity comes from a persistent monotonic
// counter so it is globally unique and never reused across resets.
export function spawnAi(
  ctx: any,
  host: any,
  difficulty: number,
  faction: Faction
): void {
  const diff = AI_PROFILES[difficulty] ? difficulty : 1;
  const cfg = ctx.db.config.id.find(0);
  const n = cfg ? cfg.nextBotId : 1n;
  if (cfg) ctx.db.config.id.update({ ...cfg, nextBotId: n + 1n });
  const botId = new Identity((1n << 255n) | n);
  foundPlayer(ctx, botId, aiName(faction, Number(n)), faction);
  ctx.db.ai.insert({
    identity: botId,
    host,
    difficulty: diff,
    decisionCd: 0,
    waveTimer: AI_PROFILES[diff].firstWaveDelay,
  });
}

// A keep just fell — its owner is out. Idempotent.
export function markDefeated(ctx: any, owner: any): void {
  const p = ctx.db.player.identity.find(owner);
  if (p && !p.defeated) ctx.db.player.identity.update({ ...p, defeated: true });
}

File: saladin-bevy/crates/protocol/src/commands/spawn.rs
use super::{building_occupancy, clamp_world, match_ctl, tech_mask_of};
use crate::components::*;
use crate::{NextEntityId, WorldConfig};
use bevy_ecs::prelude::*;
use saladin_sim::*;

/// Scatter every resource node kind across the map for `match_id` from the
/// seeded pure worldgen — call once at match start (mirrors the SpacetimeDB
/// `scatterNodes`). Reproducible: every client lays the same forest.
pub fn scatter_world_nodes(world: &mut World, match_id: u64) {
    let seed = world.resource::<WorldConfig>().seed;
    let mut nodes = scatter_nodes(seed, &node_kinds());
    // fair starts: top every spawn slot up to the guaranteed minimum of
    // wood/stone/food within FAIR_RADIUS — mirrored across players
    let extra = fair_start_nodes(seed, &nodes, MAX_PLAYERS, TREE_WOOD, STONE_YIELD, FOOD_YIELD);
    nodes.extend(extra);
    for n in nodes {
        let id = world.resource_mut::<NextEntityId>().alloc();
        world.spawn((
            GameId(id),
            MatchId(match_id),
            Pos { pos: n.pos, facing: Fx::ZERO },
            ResourceNode { res_type: n.res_type, remaining: n.yield_ },
        ));
    }
}

/// Eight fixed directions for the deterministic peasant start-cluster (the TS
/// version used cos/sin; replaced for determinism).
pub(crate) const DIRS8: [(Fx, Fx); 8] = [
    (saladin_sim::fx!("1"), saladin_sim::fx!("0")),
    (saladin_sim::fx!("0.7"), saladin_sim::fx!("0.7")),
    (saladin_sim::fx!("0"), saladin_sim::fx!("1")),
    (saladin_sim::fx!("-0.7"), saladin_sim::fx!("0.7")),
    (saladin_sim::fx!("-1"), saladin_sim::fx!("0")),
    (saladin_sim::fx!("-0.7"), saladin_sim::fx!("-0.7")),
    (saladin_sim::fx!("0"), saladin_sim::fx!("-1")),
    (saladin_sim::fx!("0.7"), saladin_sim::fx!("-0.7")),
];

pub(crate) fn spawn_building(world: &mut World, owner: u64, kind: BuildingKind, pos: V2, match_id: u64) -> u64 {
    let mask = tech_mask_of(world, owner);
    let def = effective_building_def(kind, mask);
    let id = world.resource_mut::<NextEntityId>().alloc();
    world.spawn((
        GameId(id),
        Owner(owner),
        MatchId(match_id),
        Pos { pos, facing: Fx::ZERO },
        Building { kind, hp: def.max_hp, cooldown: Fx::ZERO, rally: pos },
    ));
    id
}

#[allow(clippy::too_many_arguments)]
pub(crate) fn spawn_unit(
    world: &mut World,
    owner: u64,
    kind: UnitKind,
    pos: V2,
    match_id: u64,
    gather_state: GatherState,
    target_node: u64,
) -> u64 {
    let mask = tech_mask_of(world, owner);
    let def = effective_unit_def(kind, mask);
    let id = world.resource_mut::<NextEntityId>().alloc();
    world.spawn((
        GameId(id),
        Owner(owner),
        MatchId(match_id),
        Pos { pos, facing: Fx::ZERO },
        Unit {
            kind,
            target: pos,
            has_target: false,
            speed: def.speed,
            gather_state,
            target_node,
            carrying: 0,
            carry_type: ResourceType::Wood,
            harvest_timer: Fx::ZERO,
            hp: def.max_hp,
            attack_target: 0,
            attack_cooldown: Fx::ZERO,
            stance: Stance::Aggressive,
            morale: MORALE_MAX,
            routing: false,
            home: pos,
            garrisoned_in: 0,
            path: vec![],
            path_idx: 0,
        },
    ));
    id
}

/// Spawn a skirmish bot: found its base, then attach the `Bot` driver component.
pub(crate) fn spawn_ai(
    world: &mut World,
    player_id: u64,
    host: u64,
    difficulty: AiDifficulty,
    faction: Faction,
    match_id: u64,
) {
    let name = ai_name(faction, (player_id as usize) % 8);
    found_player(world, player_id, name, faction, match_id);
    let pe = {
        let mut q = world.query::<(Entity, &Player)>();
        q.iter(world).find(|(_, p)| p.player_id == player_id).map(|(e, _)| e)
    };
    if let Some(e) = pe {
        let prof = ai_profile(difficulty);
        world.entity_mut(e).insert(Bot {
            host,
            difficulty,
            decision_cd: Fx::ZERO,
            wave_timer: prof.first_wave_delay,
            phase: AiPhase::Boot,
            scout_id: 0,
            threat_timer: Fx::ZERO,
        });
    }
}

pub(crate) fn found_player(world: &mut World, player_id: u64, name: &str, faction: Faction, match_id: u64) {
    let seed = world.resource::<WorldConfig>().seed;
    match_ctl::ensure_match(world, match_id, player_id);

    // stable slot among players already in this match
    let used: Vec<i32> = {
        let mut q = world.query::<(&Player, &MatchId)>();
        q.iter(world).filter(|(_, m)| m.0 == match_id).map(|(p, _)| p.slot as i32).collect()
    };
    let slot = alloc_slot(&used, MAX_PLAYERS as i32).max(0);
    let keep_fp = building_def(BuildingKind::Keep).footprint;
    // validated site on the dominant landmass with open ground around it (a
    // keep against cliffs/water strands the deposit economy); the occupancy
    // pass then only needs to dodge other players' buildings
    let site = find_keep_site(seed, slot as usize, keep_fp);

    let occ = building_occupancy(world, true);
    let passable = |tx: i32, ty: i32| is_passable(seed, tx, ty) && !occ.contains(&tile_key(tx, ty));
    let base = find_buildable_near(site.x, site.y, keep_fp, passable);

    let keep_id = spawn_building(world, player_id, BuildingKind::Keep, base, match_id);

    let player_ent_id = world.resource_mut::<NextEntityId>().alloc();
    world.spawn((
        GameId(player_ent_id),
        MatchId(match_id),
        Player {
            player_id,
            name: name.to_string(),
            faction,
            stock: Stockpile { wood: START_WOOD, stone: START_STONE, food: START_FOOD, gold: START_GOLD },
            color: (slot as usize % PLAYER_COLORS.len()) as u8,
            online: true,
            keep: keep_id,
            defeated: false,
            slot: slot as u8,
            tech_mask: 0,
        },
    ));

    // nodes for gather assignment
    let nodes: Vec<(u64, V2, ResourceType)> = {
        let mut q = world.query::<(&GameId, &Pos, &ResourceNode, &MatchId)>();
        q.iter(world).filter(|(_, _, _, m)| m.0 == match_id).map(|(g, p, n, _)| (g.0, p.pos, n.res_type)).collect()
    };
    let available: Vec<ResourceType> = {
        let mut seen = Vec::new();
        for (_, _, rt) in &nodes {
            if !seen.contains(rt) {
                seen.push(*rt);
            }
        }
        seen
    };
    let types = balanced_gather_types(&available, START_PEASANTS as usize);

    for i in 0..START_PEASANTS as usize {
        let (dx, dy) = DIRS8[i % DIRS8.len()];
        let px = clamp_world(base.x + dx * SPAWN_CLUSTER);
        let py = clamp_world(base.y + dy * SPAWN_CLUSTER);
        let pos = V2::new(px, py);
        // nearest node of the assigned type
        let (state, node) = if let Some(&want) = types.get(i) {
            let typed: Vec<(u64, V2)> =
                nodes.iter().filter(|(_, _, rt)| *rt == want).map(|(id, p, _)| (*id, *p)).collect();
            let pool: Vec<(u64, V2)> =
                if typed.is_empty() { nodes.iter().map(|(id, p, _)| (*id, *p)).collect() } else { typed };
            let mut best: Option<u64> = None;
            let mut best_d = Fx::MAX;
            for (id, p) in &pool {
                if !node_reachable(seed, pos, *p) {
                    continue;
                }
                let d = dist2(pos, *p);
                if d < best_d {
                    best_d = d;
                    best = Some(*id);
                }
            }
            match best {
                Some(id) => (GatherState::ToResource, id),
                None => (GatherState::Idle, 0),
            }
        } else {
            (GatherState::Idle, 0)
        };
        spawn_unit(world, player_id, UnitKind::Peasant, pos, match_id, state, node);
    }
}
